google/cloud/storage/internal/policy_document_request.cc

```diff
@@ -56,96 +60,160 @@ nlohmann::json TransformConditions(
   return res;
 }
 
-/// If c is ASCII escape it and append it to result. Return if it is ASCII.
-bool EscapeAsciiChar(std::string& result, char32_t c) {
-  switch (c) {
+// https://en.wikipedia.org/wiki/UTF-8
+
+// The masks for 1-byte, 2-byte, 3-byte and 4-byte UTF-8 encodings. The bits
+// that are *set* in these masks are used to extract the marker bits, i.e.,
+// those indicating the length of the encoding. Negating the mask let's you
+// extract the value bits.
+auto constexpr kMask1 = 0b1000'0000U;
+auto constexpr kMask2 = 0b1110'0000U;
+auto constexpr kMask3 = 0b1111'0000U;
+auto constexpr kMask4 = 0b1111'1000U;
+
+// The mask for the trailing bytes.
+auto constexpr kMaskTrail = 0b1100'0000U;
+
+inline bool IsEncoded(char c, std::int32_t mask) {
+  return (static_cast<std::uint8_t>(c) & mask) == ((mask - 1) & mask);
+}
+
+Status ValidateUTF8Encoding(absl::string_view s, std::size_t pos,
+                            std::size_t n) {
+  // `pos` is always <= s.size().
+  if (s.size() - pos < n) {
+    return InvalidArgumentError(
+        absl::StrCat("Expected UTF-8 string, found partial UTF-8 encoding at ",
+                     pos, " string=<", s, ">"),
+        GCP_ERROR_INFO());
+  }
+  // Note that this function is always called with `n > 0`. All the call sites
```

Comment:
If `n > 0` then I assume we really want `pos < s.size()` above?

[Aside: Perhaps a single "REQUIRES" comments would be clearer for both.]

Reply:
I hope I did this right.